text/0000-default-field-values.md

```diff
@@ -0,0 +1,1658 @@
+}
+```
+
+### `proptest_derive`
+
+The yet to be released `#[derive(Arbitrary)]` custom derive macro also has a
+notion of defaults; for example, you may write:
+
+```rust
+#[derive(Debug, Arbitrary)]
+struct Foo {
+    bar: usize,
+    #[proptest(value = "42")]
+    baz: usize,
+}
+```
+
+This will generate random `usize`s for `bar` and fix the value of `baz` to `42`.
+One could instead imagine writing:
+
+```rust
+#[derive(Debug, Arbitrary)]
+struct Foo {
+    bar: usize,
+    #[proptest(default)]
+    baz: usize = 42,
+}
+```
+
+This would then fix the value of `baz` to `42`.
+
+### Conclusion
+
+As seen in the previous sections, rather than make deriving `Default`
+more magical, by allowing default field values in the language,
+user-space custom derive macros can make use of them.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+## Providing field defaults
+
+Consider a data-type such as (1):
+
+```rust
+pub struct Probability {
+    value: f32,
+}
+```
+
+You'd like encode the default probability value to be `0.5`;
+With this RFC now you can provide such a default directly where `Probability`
+is defined like so (2):
+
+```rust
+pub struct Probability {
+    value: f32 = 0.5,
+}
+```
+
+Having done this, you can now construct a `Probability` with a struct
+initializer and leave `value` out to use the default (3):
+
+```rust
+let prob = Probability { .. };
+```
+
+## Deriving `Default`
+
+Previously, you might have instead implemented the `Default` trait like so (4):
+
+```rust
+impl Default for Probability {
+    fn default() -> Self {
+        Self { value: 0.5 }
+    }
+}
+```
+
+You can now shorten this to (5):
+
+```rust
+impl Default for Probability {
+    fn default() -> Self {
+        Self { .. }
+    }
+}
+```
+
+However, since you had specified `value: f32 = 0.5` in the definition of
+`Probability`, you can take advantage of that to write the more simpler
+and more idiomatic (6):
+
+```rust
+#[derive(Default)]
+pub struct Probability {
+    value: f32 = 0.5,
+}
+```
+
+Having done this, a `Default` implementation equivalent to the one in (5)
+will be generated for you.
+
+## More fields
+
+As you saw in the [summary], you are not limited to a single field and all
+fields must not have any defaults associated with them. Instead, you can
+freely mix and match (7):
+
+```rust
+#[derive(Default)]
+struct ExprType {
+    expr: Box<Expr>,
+    ty: Box<Ty>
+    attrs: Vec<Attribute> = Vec::new(),
+}
+```
+
+Given this definition, you can omit `attrs` and `color_token` (8):
+
+```rust
+let ascribe = ExprType { expr: make_expr(), ty: make_type(), .. };
+```
+
+You can also elect to override the provided default (9):
+
+```rust
+let ascribe = ExprType {
+    attr: vec![
+        parse_quote!(#[allow(unused_parens)]),
+        parse_quote!(#[deny(other_lint)]),
+    ],
+    expr: parse_quote!((1 + 2)),
+    ty: parse_quote!(u8),
+}
+```
+
+## Default fields values are [`const` context]s
```

Comment:
Why require this? It seems like they could just be functions evaluated in the position they're used. The "constness" of a given field initializer could be inferred to determine whether a particular field's initializer is `const`. I can see arguments both ways, but the fact that I can immediately come up w/ examples of things I'd want defaults for that wouldn't be const-valid hints to me that we'll want to expand this.

Reply:
Also, this could prevent some of the macros given above from being able to remove their custom `#[default]` attributes, since many (all?) of those attributes allow non-const expressions.

Reply:
idea: warn-by-default lint against non-const defaults (at definition site, not at usage)

Reply:
Ping-- @Centril did you have a chance to think more about this?